src/nupic/algorithms/SpatialPooler.cpp

```diff
@@ -616,19 +616,36 @@ UInt SpatialPooler::countConnected_(const vector<Real> &perm) const {
 }
 
 
-UInt SpatialPooler::raisePermanencesToThreshold_(vector<Real>& perm,
-                                                 const vector<UInt>& potential) const
-{
-  clip_(perm, false);
-  UInt numConnected = countConnected_(perm);
-  while (numConnected < stimulusThreshold_) //TODO avoid the while-true loop, grow syns in 1 step
-  {
-    for (auto & elem : potential) {
-      perm[elem] += synPermBelowStimulusInc_;
-    }
-    numConnected = countConnected_(perm);
+UInt SpatialPooler::raisePermanencesToThreshold_(
+                      vector<Real>& perm,
+                      const vector<UInt>& potential) const {
+
+  if( stimulusThreshold_ == 0 )
+    return -1;
+
+  // Sort the potential pool by permanence values, and look for the synapse with
+  // the N'th greatest permanence, where N is the desired minimum number of
+  // connected synapses.  Then calculate how much to increase the N'th synapses
+  // permance by such that it becomes a connected synapse.
+
+  vector<UInt> index(potential.begin(), potential.end()); // Sort a copy, not the original!
+  auto minPermSyn = index.begin() + stimulusThreshold_ - 1;
+  // Do a partial sort, it's faster than a full sort. Only minPermSyn is in its
+  // final sorted position.
+  nth_element(index.begin(), minPermSyn, index.end(),
+      [&](UInt &A, UInt &B) { return perm[A] > perm[B]; });
+
+  Real increment = synPermConnected_ - perm[ *minPermSyn ];
```

Comment:
Const

Reply:
Also, you have to grow in multiples of the Inc constant

Reply:
I'd rather not.  I think that the increment is an artifact of the way this method was previously computed.  I don't think it will change the overall results in any substantial way.

Reply:
I’ll have to think about this. You basically disregard use of ‘synPermBelowStimulusInc’
- so you should have removed it from SP altogether 
- I’m not sure if some logic can be around synapses growing by fixed step (there are 2 ways of growth, inc on connected, and this inc on disconnected)

Reply:
@ctrl-z-9000-times

Reply:
synPermBelowStimulusInc is still used by method bumpUpWeakColumns_ so we should keep it.